### PR TITLE
Add support for Signature class instances

### PR DESCRIPTION
**Pull request by PradyumnaKrishna**
Fixes: #360

### Description of the changes being introduced by the pull request:
Currently, `verify_siganture()` method takes a signature dictionary as parameter and there is no implementation for `Signature` class present in `securesystemslib.signer`. This Pull Request adds support for `Signature` class object in `create_signature()` and `verify_signature()` methods.

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by jku** [comment]
Nice work. I think the high level issue here is that this is public API that is being used. So while I think something like this is indeed a good path forward -- in my opinion public API arguments and return values should be better defined than "it's a dict!" -- securesystemslib might not want to just change the API... But I'm just assuming: hopefully actual maintainers have opinions here.

**Comment by lukpueh** [comment]
Impeccable work, @PradyumnaKrishna. There is nothing for me to comment on in the patch. 💯 

But @jku is right that we need to consider carefully if we want to  merge (now) and thus disrupt in-toto/tuf operations in the next release.

I suggest we hold off with merging until we have a clear plan for #270.

cc @adityasaky, @SantiagoTorres

**Review by adityasaky — COMMENTED** [review]
I like the changes here, but I'm indeed wary about this breaking in-toto / python-tuf. That said, I suspect (I haven't had a chance to look) modifying both in parallel to switch away from the `dict` shouldn't be too difficult or a breaking change by any means. @PradyumnaKrishna, can I ask you to take a look at what it would mean to make that change? :)

**Comment by PradyumnaKrishna** [comment]
@adityasaky In in-toto, `Metablock` uses these functions in `sign` and `verify_signature` methods. in-toto metadata has a list of signatures, which also consist GPG Signatures in format of `dict`.
Simplest way to achieve this is by converting the Signature object to dict in `sign` function while converting the stored dict into a signature object in `verify` function of `Metablock`.

**Comment by adityasaky** [comment]
@lukpueh @SantiagoTorres thoughts on modifying in-toto simultaneously?

**Comment by lukpueh** [comment]
I gave this more thought and would like to suggest to not touch `keys.{create, verify}_signature` for the time being. Let's instead settle on a good alternative first, start using it in tuf and in-toto, and only then deprecate/refactor the current signing/verifying API.

Good alternatives:
-  [`Signer.sign`](https://github.com/secure-systems-lab/securesystemslib/blob/bd4afbe8eb2ded83fc3b1d7a81618a2a9c7242fa/securesystemslib/signer.py#L97)
- [`Key.verify_signature`](https://github.com/theupdateframework/python-tuf/blob/39f03ac8b161ad3063b995fcbd5c64b506a7dafc/tuf/api/metadata.py#L733) (currently implemented in python-tuf metadata API, but seems like a good fit for `securesystemslib`)

File: securesystemslib/keys.py [hunk 1]
```diff
@@ -803,14 +801,15 @@ def verify_signature(key_dict, signature, data):
   formats.ANYKEY_SCHEMA.check_match(key_dict)
 
   # Does 'signature' have the correct format?
-  formats.SIGNATURE_SCHEMA.check_match(signature)
+  if not isinstance(signature, signer.Signature):
+    raise exceptions.FormatError("Wanted a Signature object.")
```
Comment (MVrachev) [inline securesystemslib/keys.py:805]: Nit, feel free to ignore if you decide:
A comment I once got a while ago: `Don't add dots or exclamation marks at the end of an exception message`.

**Review by MVrachev — COMMENTED** [review]
I like the changes, good work!

**Comment by jku** [comment]
My late review and links to related work -- I believe this mostly duplicates lukas last comment:

I think modifying keys module  API is not a good idea:
  * it is a (hard) breaking change for all current users
  * keys module is not a great API as a whole, so getting it to a good place with small tweaks... is going to require a lot of small tweaks. So this would not be the last API break
  * I think the Signer + Key approach that was taken in the intoto branch (and in my related PR) are better: the create a better API on top of the current functionality. This allows us to see if the new API is good and if it is, then start deprecating and improving the implementation underneath -- requiring only one deprecation and API break. Whether signature verification is implemented in Key or Signature I don't care so much, but I do think keys module API should not change
  
Links:
* [intoto fork](https://github.com/secure-systems-lab/securesystemslib/compare/master...in-toto:securesystemslib:master): this is a combination of Key changes, signer changes and a new DSSE enveloper implementation
* My [draft PR](https://github.com/secure-systems-lab/securesystemslib/pull/456) with Key + Signer improvements, partly overlapping with the above

I think this PR should maybe  be closed?

**Pull request closed without merging**